File: built-in plugins dsp/mb-ethnic-zurna.cpp
/**
 * MB Zurna
 * Category : instrument
 * Type     : ethnic
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Turkish double-reed oboe with piercing outdoor tone
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_ETHNIC_ZURNA_H
#define MB_ETHNIC_ZURNA_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbEthnicZurna : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-ethnic-zurna";
    static constexpr const char* PLUGIN_NAME    = "MB Zurna";
    static constexpr const char* PLUGIN_TYPE    = "ethnic";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float brightness = 0.8f;  // range [0, 1]
    float reed = 0.6f;  // range [0, 1]
    float vibrato = 0.3f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbEthnicZurna() = default;
    ~MbEthnicZurna() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.reed = std::clamp(params.reed, 0f, 1f);
        params.vibrato = std::clamp(params.vibrato, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Zurna
        return input;
    }
};

#endif // MB_ETHNIC_ZURNA_H

File: maxbooster7.5/built-in plugins dsp/mb-ethnic-zurna.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbEthnicZurnaPlugin: PluginDefinition = {
  id: "mb-ethnic-zurna",
  slug: "mb-ethnic-zurna",
  name: "MB Zurna",
  category: "instrument",
  type: "ethnic" as any,
  version: "1.0.0",
  description: "Turkish double-reed oboe with piercing outdoor tone",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: 0, gain: 0.5 },
    { type: "square", detune: 3, gain: 0.3 },
    { type: "sine", detune: 1200, gain: 0.2 },
  ],
  envelope: { attack: 0.02, decay: 0.15, sustain: 0.92, release: 0.1 },
  parameters: [
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "reed",
      name: "Reed",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "vibrato",
      name: "Vibrato",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { brightness: 0.8, reed: 0.6, vibrato: 0.3, volume: 0.7 },
};

export default MbEthnicZurnaPlugin;
